<a href="https://colab.research.google.com/github/dbarrau/TensorFlow_training/blob/main/00_exercises_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 00. TensorFlow Fundamentals Exercises

In [ ]:
import tensorflow as tf
import numpy as np

!nvidia-smi


Fri Jun 10 17:43:07 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   58C    P0    28W /  70W |    444MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

### Create a vector, scalar, matrix and tensor with values of your choosing using `tf.constant()`.

Find the shape, rank and size of the tensors you created.


In [ ]:
# Create scalar

s = tf.constant(1)

# Compute rank, shape and size of s
s.ndim,s.shape, tf.size(s)


(0, TensorShape([]), <tf.Tensor: shape=(), dtype=int32, numpy=1>)

In [ ]:
# Create vector

v = tf.constant([1])

# Compute rank, shape and size of v
v.ndim, v.shape, tf.size(v)


(1, TensorShape([1]), <tf.Tensor: shape=(), dtype=int32, numpy=1>)

In [ ]:
# Create matrix

M = tf.constant([[2,4,6],
                 [1,3,5],
                 [8,10,12]])

# Compute rank, shape and size of M
M.ndim, M.shape, tf.size(M)

(2, TensorShape([3, 3]), <tf.Tensor: shape=(), dtype=int32, numpy=9>)

In [ ]:
# Create tensor
T = tf.constant([[[1,2,3],
                  [4,5,6],
                  [7,8,9]],
                 [[10,11,12],
                  [13,14,15],
                  [1,2,3]]])

# Compute rank, shape and size of M
T.ndim, T.shape, tf.size(T)

(3, TensorShape([2, 3, 3]), <tf.Tensor: shape=(), dtype=int32, numpy=18>)

### 3. Create two tensors containing random values between 0 and 1 with shape `[5, 300]`.

In [ ]:
# Set global seed
tf.random.set_seed(12)

# Instantiate T1 tensor according to requirements, with operational seed
T1 = tf.random.uniform([5,300], 0, 1, seed = 1)

# Instantiate T2 tensor according to requirements, with oprational seed
T2 = tf.random.uniform([5,300], 0, 1, seed = 2)

### 4. Multiply the two tensors using matrix multiplication

In [ ]:
# matmul operation transposing T1
tf.matmul(tf.transpose(T1),T2)

<tf.Tensor: shape=(300, 300), dtype=float32, numpy=
array([[1.1818696 , 0.732297  , 1.6789985 , ..., 0.65916395, 1.2965248 ,
        1.2094191 ],
       [1.4673159 , 1.1271098 , 2.8603518 , ..., 1.1077428 , 1.7714152 ,
        1.99566   ],
       [1.4511633 , 0.96860707, 2.2182117 , ..., 1.1956693 , 1.8527243 ,
        1.5258878 ],
       ...,
       [0.9333143 , 0.598365  , 1.4939263 , ..., 0.5114735 , 1.0597419 ,
        1.0204374 ],
       [1.2306621 , 0.7045996 , 1.6594949 , ..., 0.6687104 , 1.3610858 ,
        1.1146669 ],
       [1.1467376 , 0.88755995, 2.2075849 , ..., 1.0306494 , 1.4801182 ,
        1.5130857 ]], dtype=float32)>

### 5. Multiply the two tensors you created in 3 using dot product

In [ ]:
# Matrix multiplication using tf.tensordot

tf.tensordot(tf.transpose(T1),T2, axes = 1)

<tf.Tensor: shape=(300, 300), dtype=float32, numpy=
array([[1.1818696 , 0.732297  , 1.6789985 , ..., 0.65916395, 1.2965248 ,
        1.2094191 ],
       [1.4673159 , 1.1271098 , 2.8603518 , ..., 1.1077428 , 1.7714152 ,
        1.99566   ],
       [1.4511633 , 0.96860707, 2.2182117 , ..., 1.1956693 , 1.8527243 ,
        1.5258878 ],
       ...,
       [0.9333143 , 0.598365  , 1.4939263 , ..., 0.5114735 , 1.0597419 ,
        1.0204374 ],
       [1.2306621 , 0.7045996 , 1.6594949 , ..., 0.6687104 , 1.3610858 ,
        1.1146669 ],
       [1.1467376 , 0.88755995, 2.2075849 , ..., 1.0306494 , 1.4801182 ,
        1.5130857 ]], dtype=float32)>

### 6. Create a tensor with random values between 0 and 1 with shape `[1, 224, 224, 3]`

In [ ]:
tf.random.set_seed(5)

# Instantiate tensor with required conditions
T3 = tf.random.uniform([224,224,3], 0, 1, seed=5)

<tf.Tensor: shape=(), dtype=float32, numpy=1.7762184e-05>

### 7. Find the min and max values of the tensor you created in 6 along the first axis

In [ ]:
# Find minimum along first axis
tf.reduce_min(T3)

<tf.Tensor: shape=(), dtype=float32, numpy=1.7762184e-05>

In [ ]:
# Find maximum value of T3 along first axis
tf.reduce_max(T3)

<tf.Tensor: shape=(), dtype=float32, numpy=0.9999975>

### 8. Create a tensor with random values of shape `[1, 224, 224, 3]` then squeeze it to change the shape to `[224, 224, 3]`

In [ ]:
# Instantiate global seed
tf.random.set_seed(3)

# Instantiate tensor with random values and required shape
T4 = tf.random.uniform([1, 224, 224, 3])

# Squeeze the tensor
T4_squeezed = tf.squeeze(T4,0)
T4_squeezed

<tf.Tensor: shape=(224, 224, 3), dtype=float32, numpy=
array([[[0.13366318, 0.5678531 , 0.4963317 ],
        [0.01686525, 0.50454175, 0.74873126],
        [0.13033688, 0.33561492, 0.6519277 ],
        ...,
        [0.09551287, 0.7745788 , 0.94565904],
        [0.4585389 , 0.24478483, 0.69737494],
        [0.13486624, 0.49011993, 0.3190508 ]],

       [[0.9132737 , 0.7370925 , 0.82477975],
        [0.7902572 , 0.8107809 , 0.61262596],
        [0.54857635, 0.7577503 , 0.6478059 ],
        ...,
        [0.22832298, 0.42732084, 0.3243487 ],
        [0.1258465 , 0.9693099 , 0.72472024],
        [0.17262292, 0.5402758 , 0.01245522]],

       [[0.18029809, 0.6981578 , 0.36399138],
        [0.16166258, 0.56943595, 0.22409868],
        [0.37093806, 0.2099936 , 0.91768014],
        ...,
        [0.04689181, 0.8925816 , 0.5845498 ],
        [0.19480121, 0.04620397, 0.83770967],
        [0.12744379, 0.22989404, 0.27232206]],

       ...,

       [[0.6833439 , 0.15678465, 0.74756753],
        [0.59

### 9. Create a tensor with shape `[10]` using your own choice of values, then find the index which has the maximum value

In [ ]:
tf.random.set_seed(41)

# Instantiate numpy array
array = np.random.randint(0,10,10)

# convert array to tensor

T5 = tf.constant(array)

In [ ]:
tf.argmax(T5)

<tf.Tensor: shape=(), dtype=int64, numpy=5>

In [ ]:
# Output one-hot encoding from T5

indices = list(T5.numpy())
print(indices)
depth = len(set(indices))

tf.one_hot(indices,depth)

[4, 3, 8, 2, 0, 7, 2, 9, 1, 3]


<tf.Tensor: shape=(10, 8), dtype=float32, numpy=
array([[0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0.]], dtype=float32)>

### Tensorflow 2 quickstart for beginners

This short introduction uses Keras to:

1. Load a prebuilt dataset
2. Build a neural network machine learning model that classifies images.
3. Train this neural network
4. Evaluate the accuracy of the model

In [ ]:
print("TensorFlow Version: ", tf.__version__)

TensorFlow Version:  2.8.2


### Load dataset

We load and prepare the MNIST dataset. We will convert the sample data from integers to floating-point numbers

In [ ]:
mnist = tf.keras.datasets.mnist

(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train, X_test = X_train / 255.0, X_test / 255.0

11501568/11490434 [==============================] - 0s 0us/step


### Build a machine learning model

Build a `tf.keras.Sequential` model by stacking layers.


In [ ]:
# Instantiate Keras model
X_train.shape

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28,28)),
  tf.keras.layers.Dense(128, activation="relu"),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)                                   
])

For each example, the model returns a vector of logits or log-odds scores, one for each class.

**logits**

The vector of raw (non-normalized) predictions that a classification model generates, which is ordinarily then passed to a normalization function. If the model is solving a multi-class classification problem, logits typically become an input to the softmax function. The softmax function then generates a vector of (normalized) probabilities with one value for each possible class.

**log-odds**

The logarithm of the odds of some event. 

If the event refers to a binary probability, then odds refer to the ratio of the probability of succes (p) to the probability of failure (1 - p). for example, suppose that a given event has a 90% probability of success and a 10% probability of failure. In this case, odds is calculated as follows:

\begin{align}
Odds = \frac{p}{(1 - p)} = \frac{.9}{.1} = 9
\end{align}


In [ ]:
predictions = model(X_train[:1]).numpy()
predictions

array([[-0.28110647, -0.39127973,  0.18370417,  0.6542619 ,  0.23377874,
        -0.28707844, -0.61258674, -0.19342759,  0.7858285 , -0.01735732]],
      dtype=float32)

the `tf.nn.softmax` function converts these logits to probabilities for each class.

In [ ]:
tf.nn.softmax(predictions).numpy()

array([[0.06793112, 0.06084448, 0.10812683, 0.17309883, 0.1136791 ,
        0.06752665, 0.04876506, 0.07415617, 0.19743893, 0.08843285]],
      dtype=float32)

>**Note:** It is possible to bake the tf.nn.softmax function into the activation function for the last layer of the network. While this can make the model output more directly interpretable, this approach is discouraged as it's impossible to provide an exact and numerically stable loss calculation for all models when using a softmax output.

In [ ]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

This loss is equal to the negative log probability of the true class: The loss is zero if the model is sure of the correct class.

This untrained model gives probabilities close to random (1/10 for each class), so the initial loss should be close to `tf.math.log(1/10) ~= 2.3`

In [ ]:
loss_fn(y_train[:1], predictions).numpy()

2.6952329

Before we start training, we configure and compile the model using Keras `Model.compile`. Set the `optimizer` class to `adam`, set the `loss_fn` function defined and specify a metric to be avaluated for the model by setting the metrics parameter to `accuracy`

In [ ]:
model.compile(optimizer = 'adam',
              loss = loss_fn,
              metrics = ['accuracy'])

### Train and evaluate the model

Use the `Model.fit` method to adjust the model parameters and minimize the loss:

In [ ]:
model.fit(X_train, y_train, epochs = 5)

Epoch 1/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.3018 - accuracy: 0.9124
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1464 - accuracy: 0.9572
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1122 - accuracy: 0.9664
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0908 - accuracy: 0.9714
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0769 - accuracy: 0.9759


The `Model.evaluate` method checks the model performance, usually on a validation set or test set.

In [ ]:
model.evaluate(X_test, y_test, verbose=2)

313/313 - 1s - loss: 0.0752 - accuracy: 0.9766 - 1s/epoch - 3ms/step


[0.07519650459289551, 0.9765999913215637]

The image classifier is now trained to ~98% accuracy on this dataset. 